File: backend/src/main/java/cn/etarch/mao/notification/task/service/WebhookUrlValidator.java
package cn.etarch.mao.notification.task.service;

import cn.etarch.mao.common.exception.BusinessException;
import cn.etarch.mao.common.result.ErrorCode;
import cn.etarch.mao.notification.task.model.NotificationChannel;
import org.springframework.stereotype.Component;

import java.net.URI;

@Component
public class WebhookUrlValidator {

    public String validate(NotificationChannel channel, String rawUrl) {
        if (rawUrl == null || rawUrl.isBlank()) {
            throw new BusinessException(ErrorCode.PARAM_INVALID, "Webhook 地址不能为空");
        }
        try {
            URI uri = URI.create(rawUrl.trim());
            if (!"https".equalsIgnoreCase(uri.getScheme()) || uri.getUserInfo() != null || uri.getFragment() != null) {
                throw invalid();
            }
            if (channel == NotificationChannel.DINGTALK) {
                if (!"oapi.dingtalk.com".equalsIgnoreCase(uri.getHost())
                        || !"/robot/send".equals(uri.getPath())
                        || queryValue(uri.getRawQuery(), "access_token") == null) {
                    throw invalid();
                }
            } else if (!"open.feishu.cn".equalsIgnoreCase(uri.getHost())
                    || uri.getPath() == null
                    || !uri.getPath().matches("/open-apis/bot/v2/hook/[^/]+")
                    || uri.getRawQuery() != null) {
                throw invalid();
            }
            return uri.toASCIIString();
        } catch (BusinessException e) {
            throw e;
        } catch (Exception e) {
            throw invalid();
        }
    }

    public String mask(String rawUrl) {
        try {
            URI uri = URI.create(rawUrl);
            String value = rawUrl;
            int keep = Math.min(4, Math.max(0, value.length() - value.lastIndexOf('/') - 1));
            String suffix = keep > 0 ? value.substring(value.length() - keep) : "";
            if ("oapi.dingtalk.com".equalsIgnoreCase(uri.getHost())) {
                return "https://oapi.dingtalk.com/robot/send?access_token=****" + suffix;
            }
            return "https://open.feishu.cn/open-apis/bot/v2/hook/****" + suffix;
        } catch (Exception e) {
            return "****";
        }
    }

    private String queryValue(String query, String key) {
        if (query == null) return null;
        for (String part : query.split("&")) {
            int separator = part.indexOf('=');
            if (separator > 0 && key.equals(part.substring(0, separator))
                    && separator < part.length() - 1) {
                return part.substring(separator + 1);
            }
        }
        return null;
    }

    private BusinessException invalid() {
        return new BusinessException(ErrorCode.PARAM_INVALID, "Webhook 地址与所选通知渠道不匹配");
    }
}

File: backend-ts/src/notification/task/webhook-url-validator.ts
import { BusinessException } from '../../common/business-exception.js';
import { ErrorCode } from '../../common/error-code.js';
import type { NotificationChannel } from './types.js';

export class WebhookUrlValidator {
  validate(channel: NotificationChannel, rawUrl: string | null | undefined): string {
    if (rawUrl == null || rawUrl.trim() === '') {
      throw new BusinessException(ErrorCode.PARAM_INVALID, 'Webhook 地址不能为空');
    }
    try {
      const uri = new URL(rawUrl.trim());
      if (uri.protocol !== 'https:' || uri.username || uri.password || uri.hash) {
        throw this.invalid();
      }
      if (channel === 'DINGTALK') {
        if (uri.hostname.toLowerCase() !== 'oapi.dingtalk.com'
          || uri.pathname !== '/robot/send'
          || this.queryValue(uri.search.slice(1), 'access_token') == null) {
          throw this.invalid();
        }
      } else if (uri.hostname.toLowerCase() !== 'open.feishu.cn'
        || uri.pathname == null
        || !/^\/open-apis\/bot\/v2\/hook\/[^/]+$/.test(uri.pathname)
        || uri.search) {
        throw this.invalid();
      }
      return uri.href;
    } catch (e) {
      if (e instanceof BusinessException) {
        throw e;
      }
      throw this.invalid();
    }
  }

  mask(rawUrl: string): string {
    try {
      const uri = new URL(rawUrl);
      const value = rawUrl;
      const keep = Math.min(4, Math.max(0, value.length - value.lastIndexOf('/') - 1));
      const suffix = keep > 0 ? value.slice(value.length - keep) : '';
      if (uri.hostname.toLowerCase() === 'oapi.dingtalk.com') {
        return `https://oapi.dingtalk.com/robot/send?access_token=****${suffix}`;
      }
      return `https://open.feishu.cn/open-apis/bot/v2/hook/****${suffix}`;
    } catch {
      return '****';
    }
  }

  private queryValue(query: string, key: string): string | null {
    if (!query) {
      return null;
    }
    for (const part of query.split('&')) {
      const separator = part.indexOf('=');
      if (separator > 0 && key === part.slice(0, separator) && separator < part.length - 1) {
        return part.slice(separator + 1);
      }
    }
    return null;
  }

  private invalid(): BusinessException {
    return new BusinessException(ErrorCode.PARAM_INVALID, 'Webhook 地址与所选通知渠道不匹配');
  }
}
